src/logger.rs

```diff
@@ -0,0 +1,87 @@
+use serde_json::json;
+use tokio::fs::File;
+use tokio::io::BufWriter;
+use tokio::prelude::*;
+use tokio::sync::mpsc;
+
+use crate::goose::GooseDebug;
+use crate::GooseConfiguration;
+
+/// Logger thread, opens a log file (if configured) and waits for messages from
+/// GooseUser threads.
+pub async fn logger_main(
+    configuration: GooseConfiguration,
+    mut log_receiver: mpsc::UnboundedReceiver<Option<GooseDebug>>,
+) {
+    // Prepare an asynchronous buffered file writer for stats_log_file (if enabled).
+    let mut debug_log_file = None;
+    if !configuration.debug_log_file.is_empty() {
+        debug_log_file = match File::create(&configuration.debug_log_file).await {
+            Ok(f) => {
+                info!(
+                    "writing errors to debug_log_file: {}",
+                    &configuration.debug_log_file
+                );
+                Some(BufWriter::new(f))
+            }
+            Err(e) => {
+                error!(
+                    "failed to create debug_log_file ({}): {}",
+                    configuration.debug_log_file, e
+                );
+                std::process::exit(1);
+            }
+        }
+    }
+
+    // Loop waiting for and writing error logs from GooseUser threads.
+    loop {
+        // Wait here until a GooseUser thread sends us an error to log, or all GooseUser threads
+        // close the error log channel.
+        match log_receiver.recv().await {
+            Some(message) => {
+                match message {
+                    Some(goose_debug) => {
+                        match debug_log_file.as_mut() {
+                            Some(file) => {
+                                match file
+                                    .write(format!("{}\n", json!(goose_debug)).as_ref())
```

Comment:
Maybe it would be good to make the log-format here configurable as well? Does it already work correctly with gaggles?

Reply:
With the stats log it was semi-trivial to add multiple formats. With the debug log it's more effort, as in particular the header and footer would need special escaping. With `serde_json` this is handled automagically. For any other formats we desired, I'd need to find an equally proficient library.

The other main format we support is CSV, and there are some large shortcomings with the existing `csv` package -- in particular, it's not async (so we'd block during logging), and it takes over ownership of the file (so the underlying code would be more complex to support two different ways of writing).

Each Gaggle Worker would need to create their own local debug log currently. But that's a good point, we should throw an error if someone tries to configure this on the Manager instance as nothing would happen. I'm not sure it's worth adding another socket for sending debug logs upstream, especially considering that web pages may be quite large.

Reply:
We now throw an error if you try and start Goose as a `--manager` and also enable `--debug-log-file`, instructing you instead to enable it on the workers.

Reply:
Thx - makes sense. I mainly asked if someone would want rather "GooseClient {}" rust default debug output instead of JSON.

Of course one could always print locally, too.

Reply:
Well, `raw` format (what I called them in the other log) would be semi-trivial to add. And agreed, they _may_ be a little easier to read than json, so maybe there's value in it. I can certainly add this if you feel it's worth it (and it's okay to have a format option without csv).

Reply:
okay, added (together with tests)